src/index.js

```diff
@@ -101,44 +101,46 @@ const deduplicate = (result, dupVals) => {
         return undefined;
     }
 
-    // we will change result as a side-effect
-    const wasChanged = dupVals.some((onePackageDuplicates) => {
-        const found = findBestMatch(onePackageDuplicates, findDuplicate(resolvedResource));
+    const { value: replaceWithFirst, path: found } = searchTrie(trie, resolvedResource, sep);
 
-        if (!found) {
-            return false;
-        }
+    // found record in trie, and it's not already optimal
+    // however it's guaranteed to be "maximal"
+    if (!replaceWithFirst || found === replaceWithFirst) {
+        return undefined;
+    }
 
-        const replaceWithFirst = onePackageDuplicates[0];
-        const resolvedDup = resolvedResource.replace(found, replaceWithFirst);
+    // replacing path by alias
+    const resolvedDup = resolvedResource.replace(found, replaceWithFirst);
 
-        const lastIndex = resolvedDup.indexOf(
-            'node_modules',
-            resolvedDup.indexOf(replaceWithFirst) + replaceWithFirst.length
-        );
+    // checking that new path and the old path are pointing to the same package
+    // as long as entries in trie are derived from locations of `package.jsons`
+    // using the same trie to find the last entry
 
-        if (lastIndex !== -1) {
-            return false;
-        }
+    // TODO: this check might not be needed in optimistic mode. Not sure we need it at the build time
+    const resolvedBase = getPackageName(trie, resolvedDup);
```

Comment:
actually, this operation might be not needed. As trie is not _character_ based, but uses "folders" - it might not match the "wrong" package, like `react-dom` instead of `react`.
Probably this operation is not needed.